src/github.com/matrix-org/dendrite/syncapi/sync/requestpool.go

```diff
@@ -62,48 +62,71 @@ func (rp *RequestPool) OnIncomingSyncRequest(req *http.Request, device *authtype
-	timer := time.AfterFunc(syncReq.timeout, func() {
-		close(timeoutChan) // signal that the timeout has expired
-	})
-
-	done := make(chan util.JSONResponse)
-	go func() {
-		currentPos := rp.notifier.WaitForEvents(*syncReq)
-		// We stop the timer BEFORE calculating the response so the cpu work
-		// done to calculate the response is not timed. This stops us from
-		// doing lots of work then timing out and sending back an empty response.
-		timer.Stop()
-		syncData, err := rp.currentSyncForUser(*syncReq, currentPos)
-		var res util.JSONResponse
+	currPos := rp.notifier.CurrentPosition()
+
+	// If this is an initial sync or timeout=0 we return immediately
+	if syncReq.since == types.StreamPosition(0) || syncReq.timeout == 0 {
+		syncData, err := rp.currentSyncForUser(*syncReq, currPos)
 		if err != nil {
-			res = httputil.LogThenError(req, err)
-		} else {
-			syncData, err = rp.appendAccountData(syncData, device.UserID, *syncReq, currentPos)
-			if err != nil {
-				res = httputil.LogThenError(req, err)
-			} else {
-				res = util.JSONResponse{
-					Code: 200,
-					JSON: syncData,
-				}
-			}
+			return httputil.LogThenError(req, err)
 		}
-		done <- res
-		close(done)
-	}()
-
-	select {
-	case <-timeoutChan: // timeout fired
 		return util.JSONResponse{
 			Code: 200,
-			JSON: types.NewResponse(syncReq.since),
+			JSON: syncData,
 		}
-	case res := <-done: // received a response
-		return res
 	}
+
+	// Otherwise, we wait for the notifier to tell us if something *may* have
+	// happened. We loop in case it turns out that nothing did happen.
+
+	timeoutChan := time.After(syncReq.timeout) // case of timeout=0 is handled above
+	for {
+		select {
+		// Wait for notifier to wake us up
+		case currPos = <-rp.makeNotifyChannel(*syncReq, currPos):
+		// Or for timeout to expire
+		case <-timeoutChan:
+			return util.JSONResponse{
+				Code: 200,
+				JSON: types.NewResponse(syncReq.since),
+			}
+		// Or for the request to be cancelled
+		case <-req.Context().Done():
+			return httputil.LogThenError(req, req.Context().Err())
+		}
+
+		// Note that we don't time out during calculation of sync
+		// response. This ensures that we don't waste the hard work
+		// of calculating the sync only to get timed out before we
+		// can respond
+
+		syncData, err := rp.currentSyncForUser(*syncReq, currPos)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+		if !syncData.IsEmpty() {
+			return util.JSONResponse{
+				Code: 200,
+				JSON: syncData,
+			}
+		}
+
+	}
+}
+
+// makeNotifyChannel returns a channel that produces the current stream position
+// when there *may* be something to return to the client. Only produces a single
+// value and then closes the channel.
+func (rp *RequestPool) makeNotifyChannel(syncReq syncRequest, sincePos types.StreamPosition) chan types.StreamPosition {
+	notified := make(chan types.StreamPosition)
+
+	go (func() {
+		currentPos := rp.notifier.WaitForEvents(syncReq, sincePos)
```

Comment:
do we not need to arrange for this to be cancelled if we time out (or the request goes away)?

Reply:
Woops, was going to add a TODO/issue. This is no worse than before, and is going to be fiddly to pass cancellations into the notifier given how its implemented

Reply:
(... so I was going to punt it to a separate PR/round of work)

Reply:
+1 for punting with a todo

Reply:
though: I'm surprised it's that fiddly. Doesn't the request context help you with this?

Reply:
Ideally yes, the context should help, but I think we're using sufficiently primitives that don't work nicely with contexts (condition variables and mutexes). My hunch is that there is a saner way of doing it, but I haven't dug into it basically.